File: titanic/python/titanic.py
## Import Titanic data from spreadsheet, impute NAs with median, create title feature, separate into train and test, fit an XGB to the data, get test AUC.

## Libraries
from datetime import datetime
start = datetime.now()
import numpy
import os
import pandas
import xgboost
import random
from sklearn import metrics, datasets

## Import
titanic = pandas.read_csv("../data/titanic.csv")


## Impute NAs with Median
print titanic.isnull().apply(sum)
# PassengerId      0
# Survived         0
# Pclass           0
# Name             0
# Sex              0
# Age            177
# SibSp            0
# Parch            0
# Ticket           0
# Fare             0
# Cabin          687
# Embarked         2

def impute(col):
  if col.apply(numpy.isreal).all(axis = 0):
    value = numpy.nanmedian(col)
  else:
    value = col.mode().iloc[0]
  return col.fillna(value)


for col in titanic.columns[titanic.isnull().any(axis = 0)]:
  titanic[col] = impute(titanic[col])

print titanic.isnull().apply(sum)
# PassengerId    0
# Survived       0
# Pclass         0
# Name           0
# Sex            0
# Age            0
# SibSp          0
# Parch          0
# Ticket         0
# Fare           0
# Cabin          0
# Embarked       0


def split_to_title(x):
  return x.split(".")[0].split(",")[1].strip()

titanic["Title"] = titanic["Name"].apply(split_to_title)
titanic["Title"].replace("Mme", "Mlle", inplace = True)
for title in ["Capt", "Don", "Major"]:
  titanic["Title"].replace(title, "Sir", inplace = True)
for title in ["Dona", "Lady", "the Countess", "Jonkheer"]:
  titanic["Title"].replace(title, "Sir", inplace = True)

titanic["Title"].value_counts()
# Mr        517
# Miss      182
# Mrs       125
# Master     40
# Sir         8
# Dr          7
# Rev         6
# Mlle        3
# Col         2
# Ms          1


## Fit XGB

#### Clean Data for XGB
dep_var = titanic["Survived"]
titanic.drop(["PassengerId", "Survived", "Name", "Ticket", "Cabin"], axis = 1, inplace = True)

#### Dummyize Data for XGB
factor_vars = titanic.columns[titanic.applymap(lambda x: type(x) == str).any()]
for factor_var in factor_vars:
  dummy_vars = pandas.get_dummies(titanic[factor_var])
  titanic = pandas.concat([titanic.drop(factor_var, axis = 1), dummy_vars], axis = 1)

##### Separate Into Train, Validation, and Test for XGB
TEST_PCT = 0.1
is_train_data = [random.random() > TEST_PCT for x in range(len(titanic))]
titanic_train = titanic[is_train_data]
titanic_test = titanic[[not x for x in is_train_data]]
dep_var_train = dep_var[is_train_data]
dep_var_test = dep_var[[not x for x in is_train_data]]

VALIDATION_PCT = 0.1
is_validation_data = [random.random() > VALIDATION_PCT for x in range(len(titanic_train))]
titanic_validation = titanic_train[is_validation_data]
titanic_train = titanic_train[[not x for x in is_validation_data]]
dep_var_validation = dep_var_train[is_validation_data]
dep_var_train = dep_var_train[[not x for x in is_validation_data]]

#### Turn data into matrix for XGB
params = {"eta": 1, "max_depth": 2, "nthread": 2, "objective": "binary:logistic"}

def train(titanic_train, titanic_validation, titanic_test, params):
  titanic_train = xgboost.DMatrix(titanic_train, label = dep_var_train)
  titanic_validation = xgboost.DMatrix(titanic_validation, label = dep_var_validation)
  titanic_test = xgboost.DMatrix(titanic_test, label = dep_var_test)

  watchlist = [(titanic_train, "train"), (titanic_validation, "eval")]
  return xgboost.train(params, titanic_train, 400, watchlist)
model = train(titanic_train, titanic_validation, titanic_test, params)

## Get Test AUC
def predict(titanic_test, dep_var_test):
  titanic_test = xgboost.DMatrix(titanic_test, label = dep_var_test)
  preds = model.predict(titanic_test)
  return metrics.roc_auc_score(numpy.array(dep_var_test), preds)
print "AUC: " + str(predict(titanic_test, dep_var_test))
print "Time: " + str(datetime.now() - start)

# AUC: 0.76245210728
# Time: 0:00:00.646733

File: titanic/R/titanic.R
## Import Titanic data from spreadsheet, impute NAs with median, create title feature, separate into train and test, fit an XGB to the data, get test AUC.

## Import
timer_start <- Sys.time()
titanic <- readr::read_csv("../data/titanic.csv")

## Impute NAs with Median
sapply(titanic, function(col) sum(surveytools2::is.na_like(col)))
# PassengerId    Survived      Pclass        Name         Sex         Age
#           0           0           0           0           0         177
#       SibSp       Parch      Ticket        Fare       Cabin    Embarked
#           0           0           0           0         687           2

mode <- function(x) {
  ux <- unique(x)
  ux[which.max(tabulate(match(x, ux)))]
}

impute <- function(col) {
  value <- if (is.numeric(col)) { median(col, na.rm = TRUE) }
  else { mode(col[!surveytools2::is.na_like(col)]) }
  col[surveytools2::is.na_like(col)] <- value
  col
}

titanic <- data.frame(sapply(titanic, impute))

sapply(titanic, function(col) sum(surveytools2::is.na_like(col)))
# PassengerId    Survived      Pclass        Name         Sex         Age
#           0           0           0           0           0           0
#       SibSp       Parch      Ticket        Fare       Cabin    Embarked
#           0           0           0           0           0           0


## Create Title Feature

split_to_title <- function(x) {
  strsplit(x, split='[.,]')[[1]][2]
}
titanic$Title <- sapply(as.character(titanic$Name), split_to_title)
titanic$Title <- sub(' ', '', titanic$Title)
titanic$Title[titanic$Title %in% c("Mme", "Mlle")] <- 'Mlle'
titanic$Title[titanic$Title %in% c('Capt', 'Don', 'Major', 'Sir')] <- 'Sir'
titanic$Title[titanic$Title %in% c('Dona', 'Lady', 'the Countess', 'Jonkheer')] <- 'Lady'
titanic$Title <- factor(titanic$Title)

table(titanic$Title)
# Col     Dr   Lady Master   Miss   Mlle     Mr    Mrs     Ms    Rev    Sir
#   2      7      3     40    182      3    517    125      1      6      5



## Fit XGB

#### Clean Data for XGB
dep_var <- titanic$Survived
titanic <- dplyr::select(titanic, -PassengerId, -Survived, -Name, -Ticket, -Cabin)

for (numeric_col in c("Age", "SibSp", "Parch", "Fare")) {
  titanic[[numeric_col]] <- as.numeric(titanic[[numeric_col]])
}

#### Dummyize Data for XGB
factor_vars <- names(titanic)[sapply(titanic, is.factor)]
for (factor_var in factor_vars) {
  dummy_vars <-  data.frame(model.matrix(as.formula(paste0("~titanic$", factor_var))))
  names(dummy_vars) <- paste(factor_var, seq_along(dummy_vars), sep = "_")
  titanic <<- cbind(dplyr::select_(titanic, paste0("-", factor_var)), dummy_vars)
}


##### Separate Into Train, Validation, and Test for XGB
TEST_PCT <- 0.1
is_train_data <- (runif(NROW(titanic)) > TEST_PCT)
titanic_train <- titanic[is_train_data,]
titanic_test <- titanic[!is_train_data,]
dep_var_train <- (as.numeric(dep_var[is_train_data]) - 1)
dep_var_test <- (as.numeric(dep_var[!is_train_data]) - 1)

VALIDATION_PCT <- 0.1
is_validation_data <- (runif(NROW(titanic_train)) > VALIDATION_PCT)
titanic_validation <- titanic_train[is_validation_data,]
titanic_train <- titanic_train[!is_validation_data,]
dep_var_validation <- dep_var_train[is_validation_data]
dep_var_train <- dep_var_train[!is_validation_data]


#### Turn data into matrix for XGB
titanic_train <- xgboost::xgb.DMatrix(data = as.matrix(titanic_train), label = dep_var_train)
titanic_validation <- xgboost::xgb.DMatrix(data = as.matrix(titanic_validation), label = dep_var_validation)
titanic_test <- xgboost::xgb.DMatrix(data = as.matrix(titanic_test), label = dep_var_test)

watchlist <- list(train = titanic_train, test = titanic_validation)
model <- xgboost::xgboost(data = titanic_train,
  nrounds = 400, learning_rate = 0.05, max_depth = 2, eta = 1, objective = "binary:logistic",
  nthread = 2,
  objective = "binary:logistic", watchlist = watchlist)


## Get Test AUC
preds <- xgboost::predict(model, titanic_test)
message("AUC: ", pROC::auc(preds, dep_var_test))
timer_end <- Sys.time()
message("Time: ", timer_end - timer_start, "s")
# AUC: 0.5
# Time: 1.3565149307251s
# TODO: WTF?
